### PR TITLE
Support hex encoded secret key for `--node-key`

### PR DESCRIPTION
**Pull request by bkchr**
Adds support for reading a hex encoded secret key when being passed as
file via `--node-key`.

Fixes: https://github.com/paritytech/substrate/issues/7025

### CODE REVIEW

**Comment by tomaka** [comment]
So if the user passes `--node-key-file`, the passed file can be either hexadecimal or binary, but if they don't, then the default path can only be binary? Is that intended?

File: client/cli/src/params/node_key_params.rs [hunk 1]
```diff
@@ -124,16 +121,36 @@ fn parse_ed25519_secret(hex: &str) -> error::Result<sc_network::config::Ed25519S
...
+fn open_ed25519_secret(file: &Path) -> error::Result<sc_network::config::Ed25519Secret> {
+	let mut file = File::open(file)
+		.map_err(|e| format!("Failed to open ed25519 secret: {:?}", e))?;
+
+	let mut content = Vec::new();
+	file.read_to_end(&mut content).map_err(|e| format!("Failed to read ed25519 secret: {:?}", e))?;
+
+	let secret_key = match String::from_utf8(content.clone())
+		.ok()
+		.and_then(|s| H256::from_str(&s).ok())
```
Comment (tomaka) [inline client/cli/src/params/node_key_params.rs:140]: Can we please check `content.len()` to be either 32 or 64 instead? I don't really trust `H256::from_str` to be future-proof.

Comment (bkchr) [reply client/cli/src/params/node_key_params.rs:140]: It is under our "control" and the docs already say that they do exactly these checks: https://docs.rs/fixed-hash/0.6.1/src/fixed_hash/hash.rs.html#595

**Comment by bkchr** [comment]
> So if the user passes `--node-key-file`, the passed file can be either hexadecimal or binary, but if they don't, then the default path can only be binary? Is that intended?

Intended by this pr, yes. But if you like, I can change the default as well to be either hex or binary.

**Comment by tomaka** [comment]
I don't have a super strong opinion, but I find it surprising that the behaviour differs based on whether or not the CLI option is passed.

File: client/cli/src/params/node_key_params.rs [hunk 2]
```diff
@@ -124,16 +121,36 @@ fn parse_ed25519_secret(hex: &str) -> error::Result<sc_network::config::Ed25519S
...
+	{
+		Some(bytes) => ed25519::SecretKey::from_bytes(bytes),
+		None => ed25519::SecretKey::from_bytes(content),
```
Comment (tomaka) [inline client/cli/src/params/node_key_params.rs:143]: What about this, then?

```suggestion
		Some(bytes) => { debug_assert_eq!(content.len(), 64); ed25519::SecretKey::from_bytes(bytes) },
		None => { debug_assert_eq!(content.len(), 32); ed25519::SecretKey::from_bytes(content) },
```

File: client/cli/src/params/node_key_params.rs (hunk 1 above)
Comment (tomaka) [reply client/cli/src/params/node_key_params.rs:140]: > It is under our "control"

But we're not an omniscient collective brain.
I totally imagine in a scenario where in a year from now someone changes `H256` to automatically add 0s if the value is not long enough, or something like that, and suddenly private keys that only contain valid hexadecimal characters will parse incorrectly.

**Comment by romanb** [comment]
This situation seems convoluted. The original mistake was to suddenly start writing secret node keys to files as hexadecimal in `generate-node-key` [here](https://github.com/paritytech/substrate/pull/4954/files#diff-b7c75b1ff16d62ee6e6b49c113c1ba85R46). The [original implementation](https://github.com/paritytech/substrate/pull/4884/files#diff-b9d9b4d1aadd6f5ab78ea0d6bfe6d6b2R422) didn't do that, so this was a regression, nevermind the fact that it probably isn't a good idea to introduce a second code path that writes node keys to files, separate from the existing code in the `config` module of `sc-network`. As far as I can tell the proper solution for now would be to revert writing secret node keys as hexadecimal to files, they're only printed hexadecimal to `stdout` and in this format can be passed back to `--node-key` on the CLI. Of course eventually there should be a single code path that is responsible for reading and writing these keys, wherever it is placed.

Relates to https://github.com/paritytech/substrate/issues/7045 and #7025.

**Comment by bkchr** [comment]
@romanb thank you for the input. :) I personally don't see any harm in also supporting hex, especially as this is better human readable.

I have now, as proposed by @tomaka made the loading of the file uniform so that it always can be either hex or raw bytes.

File: client/network/src/config.rs [hunk 3]
```diff
@@ -625,10 +625,26 @@ impl NodeKeyConfig {
...
 			Ed25519(Secret::File(f)) =>
-				get_secret(f,
-					|mut b| ed25519::SecretKey::from_bytes(&mut b),
+				get_secret(
+					f,
+					|mut b| {
+						match String::from_utf8(b.to_vec())
+							.ok()
+							.and_then(|s|{
+								if s.len() == 64 {
```
Comment (bkchr) [inline client/network/src/config.rs:634]: @tomaka Okay?

**Comment by tomaka** [comment]
If I understand the problem, `subkey` started generating a file with the private key as hex. Someone who uses `subkey` to generate a file and copies it to `.local/share/...`, it wouldn't work even after this PR.
Therefore I'd be of the opinion to change [the code in `sc_network` instead](https://github.com/paritytech/substrate/blob/beb74f4923d64447935fa2ce20c2aeb4bbaaf061/client/network/src/config.rs#L638-L665).

File: client/network/src/config.rs (hunk 3 above)
Comment (tomaka) [reply client/network/src/config.rs:634]: We can even go the extra mile and use `hex::decode` instead :rocket:

**Review by tomaka — APPROVED** [review]

**Pull request merged**